component/stream/read_write_filename.go

```diff
@@ -0,0 +1,478 @@
+	if err != nil {
+		log.Err("Stream::FlushFile : error flushing file %s [%s]", options.Handle.Path, err.Error())
+		return err
+	}
+	options.Handle.Flags.Clear(handlemap.HandleFlagDirty)
+	return nil
+}
+
+// TODO: any cached file to be purged without flushing
+func (rw *ReadWriteFilenameCache) DeleteFile(options internal.DeleteFileOptions) error {
+	// log.Trace("Stream::DeleteFile : name=%s", options.Name)
+	if !rw.StreamOnly {
+		rw.purge(options.Name, false)
+	}
+	err := rw.NextComponent().DeleteFile(options)
+	if err != nil {
+		log.Err("Stream::DeleteFile : error deleting file %s [%s]", options.Name, err.Error())
+		return err
+	}
+	return nil
+}
+
+// TODO: any cached file to be purged without flushing
+func (rw *ReadWriteFilenameCache) DeleteDirectory(options internal.DeleteDirOptions) error {
+	log.Trace("Stream::DeleteDirectory : name=%s", options.Name)
+	for fileName := range rw.fileCache {
+		if strings.HasPrefix(fileName, options.Name) {
+			rw.purge(fileName, false)
+		}
+	}
+	err := rw.NextComponent().DeleteDir(options)
+	if err != nil {
+		log.Err("Stream::DeleteDirectory : error deleting directory %s [%s]", options.Name, err.Error())
+		return err
+	}
+	return nil
+}
+
+// TODO: any cached file to be purged without flushing
+func (rw *ReadWriteFilenameCache) RenameDirectory(options internal.RenameDirOptions) error {
+	log.Trace("Stream::RenameDirectory : name=%s", options.Src)
+	for fileName := range rw.fileCache {
+		if strings.HasPrefix(fileName, options.Src) {
+			rw.purge(fileName, false)
+		}
+	}
+	err := rw.NextComponent().RenameDir(options)
+	if err != nil {
+		log.Err("Stream::RenameDirectory : error renaming directory %s [%s]", options.Src, err.Error())
+		return err
+	}
+	return nil
+}
+
+// Stop : Stop the component functionality and kill all threads started
+func (rw *ReadWriteFilenameCache) Stop() error {
+	log.Trace("Stopping component : %s", rw.Name())
+	if !rw.StreamOnly {
+		rw.Lock()
+		defer rw.Unlock()
+		for fileName, buffer := range rw.fileCache {
+			delete(rw.fileCache, fileName)
+			buffer.Lock()
+			defer buffer.Unlock()
+			buffer.Purge()
+			atomic.AddInt32(&rw.CachedObjects, -1)
+		}
+	}
+	return nil
+}
+
+// GetAttr : Try to serve the request from the attribute cache, otherwise cache attributes of the path returned by next component
+func (rw *ReadWriteFilenameCache) GetAttr(options internal.GetAttrOptions) (*internal.ObjAttr, error) {
+	// log.Trace("AttrCache::GetAttr : %s", options.Name)
+	attrs, err := rw.NextComponent().GetAttr(options)
+	if err != nil {
+		log.Err("Stream::GetAttr : error getting attributes %s [%s]", options.Name, err.Error())
+		return nil, err
+	}
+	rw.RLock()
+	defer rw.RUnlock()
+	buffer, found := rw.fileCache[options.Name]
+	if !found {
+		return attrs, err
+	}
+	attrs.Mtime = buffer.Mtime
+	attrs.Size = buffer.Size
+	return attrs, nil
+}
+
+func (rw *ReadWriteFilenameCache) purge(fileName string, close bool) {
+	// check if this file is cached
+	rw.Lock()
```

Comment:
here we are locking the entire streaming layer which might hurt the performance. We might need to rethink on this.

Reply:
that's the issue with filename caching, synchronization and perf hits, if we're removing an item from the cache and another handle tries to look it up, we will have a race. Hence the lock